File: calculator.py
# python calculator.py

def readNumber(line, index): # read number
    number = 0
    while index < len(line) and line[index].isdigit():
        number = number * 10 + int(line[index])
        index += 1
    if index < len(line) and line[index] == '.':
        index += 1
        keta = 0.1
        while index < len(line) and line[index].isdigit():
            number += int(line[index]) * keta
            keta *= 0.1
            index += 1
    token = {'type': 'NUMBER', 'number': number}
    return token, index


def readPlus(line, index): # read +
    token = {'type': 'PLUS'}
    return token, index + 1


def readMinus(line, index): # read -
    token = {'type': 'MINUS'}
    return token, index + 1

def readMulti(line, index): # read *
    token = {'type': 'MULTI'}
    return token, index + 1

def readDivision(line, index): # read /
    token = {'type': 'DIVIS'}
    return token, index + 1


def tokenize(line):
    tokens = []
    index = 0
    while index < len(line):
        if line[index].isdigit():
            (token, index) = readNumber(line, index)
        elif line[index] == '+':
            (token, index) = readPlus(line, index)
        elif line[index] == '-':
            (token, index) = readMinus(line, index)
        elif line[index] == '*':
            (token, index) = readMulti(line, index)
        elif line[index] == '/':
            (token, index) = readDivision(line, index)
        else:
            print 'Invalid character found: ' + line[index]
            exit(1)
        tokens.append(token)
    return tokens


def evaluateMultiplyAndDivide(tokens): # * / calculate
    tokens2 = []

    tokens.insert(0, {'type': 'MULTI'})　# Insert a dummy '*' token
    index = 1
    number = 1

    while index < len(tokens):

        if tokens[index]['type'] == 'NUMBER':
            
            if tokens[index-1]['type'] == 'MULTI':
                number = number * tokens[index]['number']
                #print number
                
            elif tokens[index-1]['type'] == 'DIVIS':
                number = number*1.0 / tokens[index]['number']
                #print number
                
            else : # tokens[index1]['type'] in {'PLUS', 'MINUS'}
                number = tokens[index]['number']
                #print number
            
                   
        elif tokens[index]['type'] in {'PLUS', 'MINUS'}:
            
            token1 = {'type': 'NUMBER', 'number': number}
            tokens2.append(token1)
            
            token2 = {'type': tokens[index]['type']}
            tokens2.append(token2)
                      
        index += 1
            
    # index >= len(tokens) last number
    token = {'type': 'NUMBER', 'number': number}
    tokens2.append(token)
       
    return tokens2



def evaluatePlusAndMinus(tokens): # + - calculate
    answer = 0
    tokens.insert(0, {'type': 'PLUS'}) # Insert a dummy '+' token
    index = 1

    while index < len(tokens):
        
        if tokens[index]['type'] == 'NUMBER':
            
            if tokens[index - 1]['type'] == 'PLUS':
                answer += tokens[index]['number']
                
            elif tokens[index - 1]['type'] == 'MINUS':
                answer -= tokens[index]['number']
            else:
                print 'Invalid syntax'
        index += 1
    return answer


def evaluate(tokens):

    tokens2 = evaluateMultiplyAndDivide(tokens)
    answer = evaluatePlusAndMinus(tokens2)

    return answer


while True:
    print '> ',
    line = raw_input()
    tokens = tokenize(line)
    answer = evaluate(tokens)
    print "answer = %f\n" % answer

File: calculator.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

int index,index2;

struct tokens_t{
  string type; // NUMBER, PLUS, MINUS ,MULTI, DIVIS
  float number;
};

tokens_t* insert(int n, tokens_t *tokens, string str){

  tokens_t tokens2[256];
  int i = index2;
  for(i; i<n+1; i--){
    tokens2[i+1] = tokens[i];
  }
  
  tokens2[n].type = str;
  index2++;
  return tokens2;
}

tokens_t readNumber(char* line){
  float number = 0;
  tokens_t token;
  token.type = "NUMBER";
  
  while((index<strlen(line)) && (('0' <=line[index]) && (line[index] <= '9'))){
    number = number*10 + (int)(line[index]-'0');
    index++;
    if((index<strlen(line)) && (line[index] == '.')){
      index ++;
      float keta = 0.1;
      while((index<strlen(line)) && (('0' <=line[index]) && (line[index] <= '9'))){
	number = number + int(line[index]-'0')*keta;
	keta = keta*0.1;
	index++;
      }
    }
  }
  token.number = number;

  return token;
}
  
tokens_t* tokensize(char* line){
  tokens_t tokens[256];
  index = 0;
  index2 = 0;

  while(index < strlen(line)){
    if(('0' <=line[index]) && (line[index] <= '9')){
      tokens[index2] = readNumber(line);
      index2++;
    }else if(line[index] == '+'){
      tokens[index2].type = 'PLUS';
      index++;
      index2++;
    }else if(line[index] == '-'){
      tokens[index2].type = 'MINUS';
      index++;
      index2++;
    }else if(line[index] == '*'){
      tokens[index2].type = 'MULTI';
      index++;
      index2++;
    }else if(line[index] == '/'){
      tokens[index2].type = 'DIVIS';
      index++;
      index2++;
    }else{
      printf("Invalid character found: %c\n", line[index]);
      exit(1);
    }
  }

  return *tokens;

}


tokens_t* evaluteMultiAndDivide(tokens_t *tokens){

  tokens_t tokens2[256],tokens3[256];

  tokens2 = insert(0, tokens, "MULTI");

  int i, j;
  float number;
  i = 1;
  j = 0;
  number = 1.0;

  while( i < index2 ){
    if(tokens2[i].type == "NUMBER"){
      if(tokens2[i-1].type == "MULTI"){
	number = number * tokens2[i].number;
      }else if(tokens2[i-1].type == "DIVIS"){
	number = number / tokens2[i].number;
      }else{
	number = tokens2[i].number;
      }
    }else if((tokens2[i].type == "PLUS")||(tokens2[i].type == "MINUS")){
      tokens3[j].type = "NUMBER";
      tokens3[j].number = number;

      j++;

      tokens3[j].type = tokens2[j].type;
      j++;
    }
    i++;
  }

  tokens3[j].type = "NUMBER";
  tokens3[j].number = number;

  index2 = j;
  
  return tokens3;
  
}


float  evalutePlusAndMinus(tokens_t *tokens){
  float answer = 0;
  tokens_t tokens2[256];

  tokens2 = insert(0, tokens, "PLUS");
  int i = 1;

  while( i <= index2 ){
    if(tokens2[i].type == "NUMBER"){
      if(tokens2[i-1].type == "PLUS"){
	answer = answer + tokens2[i].number;
      }else if(tokens2[i-1].type == "MINUS"){
	answer = answer - tokens2[i].number;
      }else{
	printf("Invalid Syntax\n");
      }
    }
    i++;
  }

  return answer; 
  
}

float evalute(tokens_t *tokens){

  tokens_t tokens2[256];
  float answer;

  tokens2 = evaluteMultiAndDivide(tokens);
  answer = evalutePlusAndMinus(tokens2);

}

int main(){
  char line[256];
  tokens_t tokens[256];
  float answer;
  
  while (true){
    printf("\n> ");
    scanf("%s",line);
    tokens = tokenize(line);
    answer = evalute(tokens):
    printf("\nanswer = %f\n",answer);
    }
  
}
